lib/query.js

```diff
@@ -0,0 +1,206 @@
+/**
+ * Query interface for the Voxel51 Vision Services API.
+ *
+ * Copyright 2017-2018, Voxel51, Inc.<br>
+ * {@link https://voxel51.com voxel51.com}
+ *
+ * @module query
+ */
+
+'use strict';
+
+let qs = require('qs');
+
+/**
+ * Base class for API queries.
+ *
+ * Provides support for queries with fully-customizable return fields, sorting,
+ * substring searching, and record limits.
+ */
+class BaseQuery {
+  /**
+   * Creates a new BaseQuery instance.
+   *
+   * @constructor
+   * @param {string} supportedFields - an array of supported query fields
+   */
+  constructor(supportedFields) {
+    this.supportedFields_ = supportedFields;
+    this.fields = [];
+    this.search = [];
+    this.sort = null;
+    this.limit = null;
+  }
+
+  /**
+   * Adds the given field to the query.
+   *
+   * @param {string} field - a query field to add
+   * @returns {BaseQuery} the updated query instance
+   */
+  addField(field) {
+    if (this.isSupportedField_(field)) {
+      this.fields.push(field);
+    }
+    return this;
+  }
+
+  /**
+   * Adds the given fields to the query.
+   *
+   * @param {array} fields - an array of query fields to add
+   * @returns {BaseQuery} the updated query instance
+   */
+  addFields(fields) {
+    let self = this;
+    fields.forEach(function(field) {
+      self.addField(field);
+    });
+    return this;
+  }
+
+  /**
+   * Adds all supported fields to the query.
+   *
+   * @returns {BaseQuery} the updated query instance
+   */
+  addAllFields() {
+    this.addFields(this.supportedFields_);
+    return this;
+  }
+
+  /**
+   * Adds the given search to the query.
+   *
+   * @param {string} field - the query field on which to search
+   * @param {string} searchStr - the search string
+   * @returns {BaseQuery} the updated query instance
+   */
+  addSearch(field, searchStr) {
+    if (this.isSupportedField_(field)) {
+      this.search.push(`${field}:${searchStr}`);
+    }
+    return this;
+  }
+
+  /**
+   * Sets the sorting behavior of the query.
+   *
+   * @param {string} field - the field on which to sort
+   * @param {boolean} [descending=true] - whether to sort in descending order
+   * @returns {BaseQuery} the updated query instance
+   */
+  sortBy(field, descending = true) {
+    if (this.isSupportedField_(field)) {
+      this.sort = `${field}:${descending ? 'desc' : 'asc'}`;
+    }
+    return this;
+  }
+
+  /**
+   * Sets the record limit of the query.
+   *
+   * @param {number} limit - the desired record limit
+   * @returns {BaseQuery} the updated query instance
+   */
+  setLimit(limit) {
+    limit = Number.parseInt(limit);
+    if (!isNaN(limit) && limit > 0) {
+      this.limit = limit;
+    }
+    return this;
+  }
+
+  /**
+   * Converts the query instance into an object suitable for passing to the
+   * `request` library.
```

Comment:
Are we passing this to `request` or to `qs` (being a bit pedantic)?

Reply:
we use `qs` to support `toString`, but when we actually send the query, we send directly with `request`